docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,309 @@
+  - [Graduation Criteria](#graduation-criteria)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Glossary
+
+## Summary
+
+CAPZ's AzureManagedControlPlane and AzureManagedMachinePool resources expose AKS's managed cluster and agent
+pool resources for Cluster API. Currently, new features in AKS require manual changes to teach CAPZ about
+those features for users to be able to take advantage of them natively in Cluster API. As a result, there are
+several AKS features available that cannot be used from Cluster API. This proposal describes how CAPZ will
+automatically make all AKS features available on an ongoing basis with minimal maintenance.
+
+## Motivation
+
+Historically, CAPZ has exposed an opinionated subset of AKS features that are tested and known to work within
+the Cluster API ecosystem. Since then, it has become increasingly clear that new AKS features are generally
+suitable to implement in CAPZ and users are interested in having all AKS features available to them from
+Cluster API.
+
+When gaps exist in the set of features available in AKS and what CAPZ offers, users of other infrastructure
+management solutions may not be able to adopt CAPZ. If all AKS features could be used from CAPZ, this would
+not be an issue.
+
+The AKS feature set changes rapidly alongside CAPZ's users' desire to utilize those features. Because making
+new AKS features available in CAPZ requires a considerable amount of mechanical, manual effort to implement,
+review, and test, requests for new AKS features account for a large portion of the cost to maintain CAPZ.
+
+Another long-standing feature request in CAPZ is the ability to adopt existing AKS clusters into management by
+Cluster API (https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1173). Making the entire AKS
+API surface area available from CAPZ is required to enable this so existing clusters' full representations can
+be reflected in Cluster API.
+
+### Goals
+
+- Narrow the gap between the sets of features offered by AKS and CAPZ.
+- Reduce the maintenance cost of making new AKS features available from CAPZ.
+
+### Non-Goals/Future Work
+
+- Automate the features available from other Azure services besides AKS.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a managed cluster user, I want to be able to use all available AKS features natively from CAPZ so that I
+can more consistently manage my CAPZ AKS clusters that use advanced or niche features without having to wait
+for each of them to be implemented in CAPZ.
+
+#### Story 2
+
+As an AKS user looking to adopt Cluster API over an existing infrastructure managed solution, I want to be
+able to use all AKS features natively from CAPZ so that I can adopt Cluster API with the confidence that all
+the AKS features I currently utilize are still supported.
+
```

Comment:
Is there a user story for new features arriving in AKS over the lifecycle of a cluster? Like

As an AKS user leveraging CAPZ to maintain my AKS fleet I want the introduction of new AKS features to automatically be available to all my existing clusters as they are made available in the AKS managed service.

(there are some AKS features that are not made available to existing AKS clusters, so we're not trying to address that, just the features that AKS rolls out and enables to the existing clusters in prod)

Reply:
I think it's still possible today to start using new AKS features in existing clusters once CAPZ is upgraded to a version which implements those features. Do you mean specifically without upgrading CAPZ to a new version? In that case an ASO upgrade would still be necessary to get the new API versions with the features. And since CAPZ installations are still tightly coupled with a particular version of ASO, I wonder if we're still basically at that "you need to upgrade CAPZ to get new AKS features" stage.

Reply:
> Do you mean specifically without upgrading CAPZ to a new version?

Yes. Because that's how it works in AKS (the managed service gets your existing cluster the new functionality, where appropriate, automatically).

So I wonder if it's worth considering this and what that would require (auto-update of CAPZ + ASO)...

Reply:
I wonder if the best we can do within the scope of this proposal is prefer solutions that don't tie one version of CAPZ to one version of the ASO APIs. Then that would open the door BYO ASO (#3900) and some hypothetical auto-updating ASO to close that loop.